Provide cross-compiling guidance when Apple SDK is missing

madsmtm opened the pull request:
Emit more guided warnings when cross-compiling to Apple platforms from non-host macOS.

Note that in the common cross-compilation case (to macOS while linking with `cc`-like compiler), users won't actually hit this, since `xcrun` isn't invoked in that case. But I intend to change that in https://github.com/rust-lang/rust/pull/131477, and then the error message here will start to matter a lot more.

The message now looks something like:
```console
$ rustc +stage1 example.rs --target aarch64-apple-ios
warning: invoking `"xcrun" "--sdk" "iphoneos" "--show-sdk-path"` to find iPhoneOS.sdk failed: No such file or directory (os error 2)
  |
  = note: the SDK is needed by the linker to know where to find symbols in system libraries and for embedding the SDK version in the final object file
  = help: pass the path to the SDK using the SDKROOT environment variable
  = help: the SDK can be downloaded and extracted from https://developer.apple.com/download/all/?q=xcode (requires an Apple ID).
          
          The full Xcode bundle should contain the SDK in Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk.
  = warning: you will also need to use a linker capable of linking Mach-O files, consider using the bundled `lld` with `-Clinker=rust-lld`
  = warning: cross-compiling on iOS, tvOS, visionOS or watchOS (in particular the linking step) is ill supported on non-macOS hosts
```

Suggestions on how to improve it welcome!

Might also make sense to emit some of these help messages if the linker invocation itself fails? Unsure, but I'll work on that in a different PR.

(I originally considered not even attempting to invoke `xcrun`, but decided to continue doing that, since an `xcrun`-compatible interface _can_ be present on other platforms, it's just unusual (e.g. Facebook developed [`xcbuild`](https://github.com/facebookarchive/xcbuild) for a while, similar probably exist)).

Follow-up to https://github.com/rust-lang/rust/pull/139010.
Part of https://github.com/rust-lang/rust/issues/129432.
Related to https://github.com/rust-lang/rustup/issues/1483 in that we now document the steps required for cross-compilation in the error message.

@rustbot label O-apple
r? wesleywiser since you just reviewed rust-lang/rust#139010
CC @BlackHoleFox @thomcc

madsmtm commented:
Converted to draft, because I want to think a little bit if we should make it a warning on host macOS too (such that linking with e.g. `zig cc` on macOS, while not having Xcode installed, still works).

wesleywiser reviewed (approved):
As written, this looks fine to me so please ping me when you move it out of draft state (whichever way you decide) 🙂

Mark-Simulacrum commented:
We might want to be a bit careful around recommending usage of the Apple SDK if our host triple isn't an apple triple - IIRC, the license terms don't permit usage of some parts of the SDK on non-apple hardware?

madsmtm commented:
Ah, yeah, [the Xcode SLA](https://www.apple.com/legal/sla/docs/xcode.pdf) does say in section "2.2 A Permitted Uses and Restrictions":
> Install a reasonable number of copies of the Apple Software [including the Apple SDKs] on Apple-branded computers that are owned or controlled by You to be used internally by You or Your Authorized Developers

And in section "2.7 Restrictions; No Other Permitted Uses":
> The grants set forth in this Agreement do not permit You to, and You agree not to, install, use or run the Apple Software [including the Apple SDKs] or Apple Services on any non-Apple-branded computer or device, or to enable others to do so.

Btw, `zig cc` is clearly already violating that, since they bundle `libSystem.tbd` and headers from the macOS SDK, and thereby "enable others to [cross compile]"...

I can change the wording to state that what the user is trying to do (cross link) is (probably, IANAL) not allowed according to Apple's terms? Though I still feel it important to provide these instructions in some shape or form.

---

Minutiae:
- `codegen_ssa_xcrun_cross_about`: technical description ("this is what an SDK is").
- `codegen_ssa_xcrun_cross_env_var`: also technical description ("this is how you pass SDKs to `rustc`").
- `codegen_ssa_xcrun_cross_download_sdk`: describes how the SDKs can be legally obtained.
  Note that I explicitly also chose to not mention that there are many ways of obtaining the SDKs without an Apple developer account.
  This way, I was hoping to defer the decision to the user, as they're obligated to read the Xcode SLA before using Xcode.
  But I could state explicitly here that doing this on non-Apple platforms is likely a violation of the SLA.
- `codegen_ssa_xcrun_cross_linker_not_explicitly_set`: technical recommendation ("here's an example of a linker you can easily and legally use on your platform").
- `codegen_ssa_xcrun_cross_ill_supported_target`: not sure, I'm really documenting the state of affairs of the linkers themselves (`zig cc` or `lld`), `rustc`'s support is as good as for macOS. Should maybe just remove this.

madsmtm commented:
> It seems like, today, if I want to produce e.g. a dylib with Rust's standard library/core but no C code, I don't need the Apple SDK to do so -- lld is more or less happy to produce a Mach-O shared object (or non-shared object) from a pure Rust crate.

Hmm, perhaps I haven't been clear enough here. The Apple SDKs provides two things: headers and linker stubs.

`lld` is _not_ able to produce neither a dylib nor binary without the linker stubs being available.

`zig cc` can in some cases because it bundles the linker stubs for system libraries (e.g. `libSystem.tdb`, the linker stub for `libSystem.dylib`, as well as `libc.tdb` etc.), but not system frameworks (`CoreFoundation.framework/CoreFoundation.tdb` etc.).

Mark-Simulacrum commented:
> lld is not able to produce neither a dylib nor binary without the linker stubs being available.

This seems untrue to me -- e.g., I produced this:

```
$ file libt.dylib
libt.dylib: Mach-O 64-bit arm64 dynamically linked shared library, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|NO_REEXPORTED_DYLIBS|HAS_TLV_DESCRIPTORS>
```

Without linking any shared libraries on a x86_64-unknown-linux-gnu host with:

```
mkdir -p /tmp/sdk
SDKROOT=/tmp/sdk rustc +nightly t.rs --target=aarch64-apple-darwin --crate-type=dylib -Zunstable-options '-Clink-arg=-undefined dynamic_lookup' -Clinker=$HOME/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/gcc-ld/ld64.lld -Clinker-flavor=darwin-lld --verbose -Csave-temps
```

That fails with:

```
  = note: rust-lld: error: unknown argument '-flavor'
          rust-lld: error: unknown argument '-undefined dynamic_lookup'
```

But, if I edit that command to:

* Remove `-flavor`
* Fix `-undefined ...` to be two arguments, not one
* Remove `-lSystem`, `-lc`, `-lm`

That produces a libt.dylib that looks functional to me. Copying that to my macOS laptop, I can then link it with C and run it:

```
$ cat test.c
void rust_shared_foo();

int main() {
    rust_shared_foo();
    return 0;
}
$ clang test.c -lt -L. -o test
$ ./test
hello world from rust
```

Unless lld has bundled something (or maybe Rust has), I didn't use zig or similar here and AFAIK there's no linker stubs involved in producing this result.

madsmtm committed:
Provide cross-compiling guidance when Apple SDK is missing

The SDK, along with a cross-linker, are the two main components required
to link pure-Rust applications from non-macOS systems, but this can be
hard to figure out.

madsmtm commented:
I talked with some Rust Foundation people back in late August (email and a call), they were busy at the time with RustConf IIRC. I've sent a follow-up email now.

madsmtm commented:
> I actually want to propose that rustc bundle a copy of `libSystem.tbd`

I've been considering the same thing. A semi-alternative would be https://github.com/rust-lang/rust/pull/146357.

One wrinkle is that this naturally forces `rustc` to keep it up-to-date with the latest SDK release (and its presence might affect the `version` and `compatibility_version` that's embedded in the final dylib, haven't tested that enough), not yet sure how to resolve that. We'd also probably need flags to control the SDK version used in the final binary (it has concrete effects, the best example I know of is that [`-[NSView wantsBestResolutionOpenGLSurface]`](https://developer.apple.com/documentation/appkit/nsview/wantsbestresolutionopenglsurface?language=objc) changes behaviour depending on the SDK version).

> Combined with a cross-linker like lld, this appears to be the _only_ file you need to successfully cross-build for macOS on Linux.

Assuming you don't use [macOS-specific frameworks](https://github.com/madsmtm/objc2), then yes.

> You don't even need the C headers because Rust binds to libc at the ABI level.

See https://github.com/rust-lang/cc-rs/issues/1585 for ideas on that.

> So this would probably be a slightly different legal question. Instead of "Is it okay to suggest that users could download the whole SDK when they're likely on a system that you're not allowed to do that on," we're asking "is it okay for us to ship this one file which is derived from an APSL source but reaches us through a Mac, just like Zig has been doing for 4+ years".

Not just _slightly_, it's significantly different.

In the email I sent to the Rust Foundation I was trying to get clarification on how much of `objc2` is allowed at the same time, that's probably why it stalled out TBH.

Feel free to send them an email yourself, that might help move things along.

geofft commented:
> The thing that influences this particular case is actually the SDK part of the `-platform_version` linker flag, which is passed automatically by Clang by reading it from the current Xcode SDK.

Aha, thanks, I can see the difference with `-Wl,-platform_version,macos,10.14,10.14` vs. `-Wl,-platform_version,macos,10.15,10.15`. Interestingly this produces more of a diff in the binary than just `LC_BUILD_VERSION`, but I can also see the difference from manually changing the `LC_BUILD_VERSION` alone and nothing else.

> Prooobably not, though I'm still wary of overriding the versions from the SDK (e.g. if users are intentionally using an older Xcode to build their applications, it'd be surprising if we started making their app behave as-if it was built using a newer Xcode).

To be clear, I'm only proposing that we implicitly use a bundled libSystem.tbd if _both_ we're not on macOS / are cross-compiling _and_ SDKROOT isn't set. If you're on macOS and you're compiling software, there's little reason not to install the SDK the normal way, especially as you're going to need stuff like a linker anyway. (But we can make it available as an option.) If you're setting SDKROOT we should respect that. So this should only take effect in the case where you do not have an existing SDK available to override and have no obvious better (slash legal) way of getting one.

Also, if it is just `LC_BUILD_VERSION` and not something in the tbd file, then that's much easier to expose as an option if we really want to / it means that updates to the tbd file itself don't implicate behavior changes.

madsmtm commented:
Update: I've sent a new email to the Rust Foundation, asking about just this PR (instead of asking two questions at the same time, with the other question being much harder).
